compiler/src/dotty/tools/dotc/cc/Capability.scala

```diff
@@ -91,33 +91,33 @@ object Capabilities:
    */
   case class Maybe(underlying: Capability) extends DerivedCapability
 
-  /** The readonly capability `x.rd`. We have {x.rd} <: {x}.
-   *
-   *  Read-only capabilities cannot wrap maybe capabilities
-   *  but they can wrap reach capabilities. We have
-   *      (x?).readOnly = (x.rd)?
-   */
-  case class ReadOnly(underlying: ObjectCapability | RootCapability | Reach | Restricted)
-  extends DerivedCapability
-
   /** The restricted capability `x.only[C]`. We have {x.only[C]} <: {x}.
    *
-   *  Restricted capabilities cannot wrap maybe capabilities or read-only capabilities
+   *  Restricted capabilities cannot wrap maybe capabilities
    *  but they can wrap reach capabilities. We have
    *      (x?).restrict[T] = (x.restrict[T])?
    *      (x.rd).restrict[T] = (x.restrict[T]).rd
    */
-  case class Restricted(underlying: ObjectCapability | RootCapability | Reach, cls: ClassSymbol)
+  case class Restricted(underlying: CoreCapability | RootCapability | Reach, cls: ClassSymbol)
   extends DerivedCapability
 
+  /** An extractor for the read-only capability `x.rd`. `x.rd` is represented as
+   *  `c.only[caps.Read]`.
+   */
+  object ReadOnly:
+    def apply(underlying: CoreCapability | RootCapability | Reach | Restricted)(using Context): Restricted =
+      Restricted(underlying.stripRestricted.asInstanceOf, defn.Caps_Read)
```

Comment:
Is it valid here to strip the existing restricted? e.g. if the classifiers was extended with a sub-classifier of `Read`

Reply:
After a second thought I think it should be fine. This expressiveness issue could be solved by first widening m.only[Control] to just async and then take readonly on it.